### PR TITLE
Constraints refactored to be more type safe

### PR DESCRIPTION
**Pull request by hcorg**
Hopefully it will ease listing test cases

### CODE REVIEW

File: src/lib/data/constraints/logicoperators.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        , m_rightChild(std::move(rightChild))
+    {}
+    ~AndConstraint() override = default;
+
+    const Constraint<ValueType> &leftChild() const { return *m_leftChild; }
+    const Constraint<ValueType> &rightChild() const { return *m_rightChild; }
+
+    void accept(ConstraintVisitor<ValueType> &visitor) const { visitor.visit(*this); }
+
+    /*    RangeList<typename RangeConstraint<T>::ValueType> asRangeList() const override
```
Comment (psockoN7) [inline src/lib/data/constraints/logicoperators.h]: seems forgotten

File: src/lib/data/constraints/logicoperators.h [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    {}
+    ~OrConstraint() override = default;
+
+    const Constraint<ValueType> &leftChild() const { return *m_leftChild; }
+    const Constraint<ValueType> &rightChild() const { return *m_rightChild; }
+
+    void accept(ConstraintVisitor<ValueType> &visitor) const { visitor.visit(*this); }
+
+    /*    RangeList<typename RangeConstraint<T>::ValueType> asRangeList() const override
+    {
```
Comment (psockoN7) [inline src/lib/data/constraints/logicoperators.h]: again, seems forgotten

File: src/lib/data/constraints/logicoperators.h (hunk 1 above)
Comment (hcorg) [reply src/lib/data/constraints/logicoperators.h]: true, thx

File: src/lib/data/constraints/logicoperators.h (hunk 2 above)
Comment (hcorg) [reply src/lib/data/constraints/logicoperators.h]: yep, fixed, thx

File: src/lib/data/constraints/rangeconstraint.h [hunk 3]
```diff
@@ -25,49 +25,40 @@
...
+    ~RangeConstraint() override = default;
 
-    const RangesTree &rangesTree() const { return m_rangesTree; }
+    const Range<typename ValueType::Type> &range() const { return m_range; }
 
-    void appendSubtree(const ExpressionTree::ExpressionNode *node)
+    void accept(ConstraintVisitor<ValueType> &visitor) const override { visitor.visit(*this); }
+
+    /*RangeList<typename RangeConstraint<T>::ValueType> asRangeList() const TODO override
+    {
```
Comment (psockoN7) [inline src/lib/data/constraints/rangeconstraint.h]: this is forgotten

Comment (hcorg) [reply src/lib/data/constraints/rangeconstraint.h]: damn, I really thought I cleaned that up :)

**Pull request merged**